pkg/utils/types/time_series.go

```diff
@@ -0,0 +1,83 @@
+package types
+
+import (
+	"sync"
+)
+
+// Timestamp is a regulation of time types
+type Timestamp int
+
+// State represent a change of value - the new value and the time of the change
+type State[T any] struct {
+	StartTime Timestamp
+	Val       T
+}
+
+// TimeSeries represents a value that changes over time.
+// It exposes the value at a given time, assuming that between two changes the value is constant.
+// The object is designed to be thread-safe.
+type TimeSeries[T any] struct {
+	states []State[T]
+	mu     sync.RWMutex
+}
+
+func NewTimeSeries[T any](defaultVal T) *TimeSeries[T] {
+	defaultState := State[T]{
+		StartTime: 0,
+		Val:       defaultVal,
+	}
+	return &TimeSeries[T]{
+		states: []State[T]{defaultState},
+	}
+}
+
+// AddState add a change in the value - from given time until next change, the value is the new one.
+func (co *TimeSeries[T]) AddState(newState State[T]) {
+	co.mu.Lock()
+	defer co.mu.Unlock()
+	insertIndex := co.getRelevantStateIndex(newState.StartTime) + 1
```

Comment:
You're always assuming that the `StartTime` of the new state is bigger than the of the last state. Should you try to guarantee that (and possibly return an error ?).

Reply:
You are correct. the Starttime is always increasing number over time and you're adding at specific indexes.